### PR TITLE
Add ripple warp mode to LaserPointer

### PR DESCRIPTION
**Pull request by kyle-emmerich**
This mode is still behind a flag

### CODE REVIEW

File: CoreScriptsRoot/Modules/VR/LaserPointer.lua [hunk 1]
```diff
@@ -876,13 +1010,25 @@ do --Laser/teleport functions
...
 		self:updateTeleportPlop(hitPoint, hitNormal)
-		self:setTeleportValidState(self:canTeleportTo(workspace.CurrentCamera.CFrame.p, self.teleportPart, self.teleportPoint, self.teleportNormal))
+		self:setTeleportValidState(self:canTeleportTo(workspace.CurrentCamera.CFrame.p, self.teleportPart, self.teleportPoint, self.teleportNormal, self.rippleHitPoint))
+	end
+
+	function LaserPointer:updateRipple()
+		local rippleOrigin = self:getRippleOriginPosition()
+		local rippleRadius = self:calculateRippleRadius()
+
+		local scaleFactor = 256 / 248
```
Comment (SolarCrane) [inline CoreScriptsRoot/Modules/VR/LaserPointer.lua]: This seems like a magic number, can we put it in some sort of settings config?

File: CoreScriptsRoot/Modules/VR/LaserPointer.lua [hunk 2]
```diff
@@ -375,6 +432,76 @@ do --Helper functions
...
+	end
+
+	function LaserPointer:calculateParabolaRipplePlaneIntersection()
+		-- x = (-b + sqrt(b^2 - 4ac)) / 2a
+		local parabolaOriginCF = self.originPart:GetRenderCFrame()
+		local planeHeight = self:getRippleOriginPosition().Y
+		local a, b, c = self.parabola.A, -self.parabola.B, parabolaOriginCF.p.Y - planeHeight
+
+		if a == 0 then
+			a = 1e-6
```
Comment (SolarCrane) [inline CoreScriptsRoot/Modules/VR/LaserPointer.lua]: 1e-6 seems to be going around a lot here, do we want to put this in some constant?

File: CoreScriptsRoot/Modules/VR/LaserPointer.lua [hunk 3]
```diff
@@ -375,6 +432,76 @@ do --Helper functions
...
+	function LaserPointer:getRippleOriginPosition()
+		local humanoid = getLocalHumanoid()
+		if not humanoid then return end
+
+		local rootPart = humanoid.Torso
+		if not rootPart then return end
+
+		local hipHeight = humanoid.HipHeight
+		if humanoid.RigType == Enum.HumanoidRigType.R6 then
+			hipHeight = 3
```
Comment (SolarCrane) [inline CoreScriptsRoot/Modules/VR/LaserPointer.lua]: I see, is there a way to fix it on the C++ side? Borked APIs are lame!

Comment (kyle-emmerich) [inline CoreScriptsRoot/Modules/VR/LaserPointer.lua]: So I talked to gameplay team, it sounds like this is going to have to stay. HipHeight isn't necessarily intended for this use, and I'm going to propose an API addition for Humanoid.GroundOffset that'll replace this code here.

**Comment by hockeychaos** [comment]
is this still going to be a thing?

**Comment by hockeychaos** [comment]
@kyle-emmerich @SolarCrane help

**Comment by kyle-emmerich** [comment]
Some day! I need to look into it again when I get time.

**Pull request closed without merging**